### ISSUE TITLE
redirects 

### ISSUE THREAD

**Issue by nimasmi** [torchbox/wagtail#2281]
In [RedirectMiddleware.process_response](https://github.com/torchbox/wagtail/blob/master/wagtail/wagtailredirects/middleware.py#L25) it is possible to encounter a `MultipleObjectsReturned` exception for a request, when there are almost-duplicate redirects, one matching a specific site, and one with `site=None`. This is because it calls `get_for_site`, which ultimately translates to:

```python
Redirect.objects.filter(models.Q(site=site) | models.Q(site=None)).get(old_path='/path')
```

I suggest changing it to something like:

```python
        # Find redirect
        try:
            redirect = models.Redirect.get(site=request.site, old_path=path)
        except models.Redirect.MultipleObjectsReturned:
            redirect = models.Redirect.get(site=None, old_path=path)
        except models.Redirect.DoesNotExist:
            if path == path_without_query:
                # don't try again if we know we will get the same response
                return response
```
i.e. preferring the specific redirect over the generic one, where possible, and ditching use of `get_for_site` (here at least)

**Comment by timheap** [comment]
I like the idea, but the provided code has a bug. `models.Redirect.get(site=request.site, old_path=path)` will never raise the `MultipleObjectsReturned` error from the old code, as it can not select a Redirect where `site=None`. This should probably still be `models.Redirect.get_for_site(request.site).get(old_path=path)`, as it is in the current code. The extra `except models.Redirect.MultipleObjectsReturned:` block will then work as intended.

**Comment by nimasmi** [comment]
Just keeping you on your toes. Updated version, then:

``` python
# Find redirect
        try:
            redirect = models.Redirect.get_for_site(request.site).get(old_path=path)
        except models.Redirect.MultipleObjectsReturned:
            redirect = models.Redirect.get(site=None, old_path=path)
        except models.Redirect.DoesNotExist:
            if path == path_without_query:
                # don't try again if we know we will get the same response
                return response
```

**Comment by timheap** [comment]
Looks good. Could you make this in to a pull requests, including a test for this case with two sites? I would be happy to merge this in if you did that.

**Comment by gasman** [comment]
Fixed in #2495
